### PR TITLE
DM-21785: Create HTM index, nDiaSources, and DiaObjectFlags plugin

### PR DESCRIPTION
**Pull request by morriscb**

### CODE REVIEW

File: tests/test_dia_calculation_plugins.py [hunk 1]
```diff
@@ -53,39 +56,149 @@ def testCalculate(self):
...
+        """
+        # Test expected pixelIds.
+        diaObject = dict()
+        diaObject["ra"] = 0.
+        diaObject["decl"] = 0.
+        plug = HTMIndexDiaPosition(HTMIndexDiaPositionConfig(),
+                                   "ap_HTMIndex",
+                                   None)
+        plug.calculate(diaObject)
+        self.assertAlmostEqual(diaObject["pixelId"], 131072)
```
Comment (mrawls) [inline tests/test_dia_calculation_plugins.py]: Can you provide any context about why this (and the subsequent example) is the correct pixelId value?

File: python/lsst/ap/association/diaCalculationPlugins.py [hunk 2]
```diff
@@ -84,6 +88,106 @@ def calculate(self, diaObject, diaSources, **kwargs):
...
             diaObject["ra"] = aveCoord.getRa().asDegrees()
             diaObject["decl"] = aveCoord.getDec().asDegrees()
+            diaObject["radecTai"] = np.max(diaSources["midPointTai"])
+
+
+class HTMIndexDiaPositionConfig(DiaObjectCalculationPluginConfig):
+
+    indexer = IndexerRegistry.makeField(
+        doc='Select the spatial indexer to use within the database.',
+        default='HTM'
```
Comment (mrawls) [inline python/lsst/ap/association/diaCalculationPlugins.py]: While HTM is a fine default here, you've hardwired "HTM" into the plugin names here. Are there situations where spatial indexers other than HTM might be desired? If so, could you call this something more generic like `SpatialIndexDiaPosition`?

File: tests/test_dia_calculation_plugins.py [hunk 3]
```diff
@@ -53,39 +56,149 @@ def testCalculate(self):
...
         self.assertAlmostEqual(diaObject["decl"], 0.0)
+        self.assertEqual(diaObject["radecTai"], 10)
 
         diaObject = dict()
         diaSources = pd.DataFrame(data={"ra": np.zeros(n_sources),
-                                        "decl": np.linspace(-1, 1, n_sources)})
+                                        "decl": np.linspace(-1, 1, n_sources),
+                                        "midPointTai": np.linspace(0,
+                                                                   n_sources,
+                                                                   n_sources)})
```
Comment (mrawls) [inline tests/test_dia_calculation_plugins.py]: I'm not sure if these are reasonable timestamps or not, are the units in MJD? Starting at 2450000-whatever might more accurately approximate reality.

File: python/lsst/ap/association/diaCalculationPlugins.py (hunk 2 above)
Comment (morriscb) [reply python/lsst/ap/association/diaCalculationPlugins.py]: That would be done by a different plugin.

File: tests/test_dia_calculation_plugins.py (hunk 3 above)
Comment (morriscb) [reply tests/test_dia_calculation_plugins.py]: It doesn't matter what the values are. Only that they are round tripped.

File: tests/test_dia_calculation_plugins.py (hunk 1 above)
Comment (morriscb) [reply tests/test_dia_calculation_plugins.py]: These are the values produced by the HTM pixelization at the locations requested. The specific values here aren't particularly important, only that they are stable and round trip correctly.

**Comment by morriscb** [comment]
> Please be a little more descriptive in your commit message(s), and consider moving the changes to `MeanDiaPosition` onto a separate commit.
> Are each plugin's output columns documented anywhere? For example, seeing the new `radecTai` output column for `MeanDiaPositionConfig` makes me wonder what the units are and why this name was chosen. It looks like a timestamp, presumably using the TAI standard, which corresponds to each DiaSource, but the name suggests the time itself is somehow a function of the ra and dec, which I don't think is what you mean.

The output columns are described in https://github.com/lsst/dax_ppdb/blob/master/data/ppdb-schema.yaml and https://github.com/lsst/cat.

File: tests/test_dia_calculation_plugins.py (hunk 3 above)
Comment (mrawls) [reply tests/test_dia_calculation_plugins.py]: I guess they won't ever be anywhere near too big to store/retrieve properly, so OK.

File: python/lsst/ap/association/diaCalculationPlugins.py (hunk 2 above)
Comment (mrawls) [reply python/lsst/ap/association/diaCalculationPlugins.py]: OK, it's a little odd to me then that you have a config option to select the spatial indexer, which implies a different choice could be made. But I don't have a suggestion for a better way to do it. Maybe a check to make sure `indexer == 'HTM'` ?

**Pull request merged**